airbyte-integrations/connectors/source-close-com/source_close_com/source.py

```diff
@@ -0,0 +1,706 @@
+    primary_key: str = "id"
+    number_of_items_per_page: int = 100
+
+    def __init__(self, **kwargs: Mapping[str, Any]):
+        super().__init__(authenticator=kwargs["authenticator"])
+        self.config: Mapping[str, Any] = kwargs
+        self.start_date: str = kwargs["start_date"]
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        decoded_response = response.json()
+        has_more = bool(decoded_response.get("has_more", None))
+        data = decoded_response.get("data", [])
+        if has_more and data:
+            parsed = dict(parse_qsl(urlparse(response.url).query))
+            # close.com has default skip param - 0. Used for pagination
+            skip = parsed.get("_skip", 0)
+            limit = parsed.get("_limit", self.number_of_items_per_page)
+            return {"_skip": int(skip) + int(limit)}
+        return None
+
+    def request_params(
+        self,
+        stream_state: Mapping[str, Any],
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> MutableMapping[str, Any]:
+
+        params = {"_limit": self.number_of_items_per_page}
+
+        # Handle pagination by inserting the next page's token in the request parameters
+        if next_page_token:
+            params.update(next_page_token)
+
+        return params
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        yield from response.json()["data"]
+
+    def backoff_time(self, response: requests.Response) -> Optional[float]:
+        """This method is called if we run into the rate limit.
+        Close.com puts the retry time in the `rate_reset` response body so
+        we return that value. If the response is anything other than a 429 (e.g: 5XX)
+        fall back on default retry behavior.
+        Rate-reset is the same as retry-after.
+        Rate Limits Docs: https://developer.close.com/#ratelimits"""
+
+        backoff_time = None
+        error = response.json().get("error", backoff_time)
+        if error:
+            backoff_time = error.get("rate_reset", backoff_time)
+        return backoff_time
+
+
+class IncrementalCloseComStream(CloseComStream):
+
+    cursor_field = "date_updated"
+    state_checkpoint_interval = 100
+
+    def get_updated_state(
+        self,
+        current_stream_state: MutableMapping[str, Any],
+        latest_record: Mapping[str, Any],
+    ) -> Mapping[str, Any]:
+        """
+        Update the state value, default CDK method.
+        For example, cursor_field can be "date_updated" or "date_created".
+        """
+        if not current_stream_state:
+            current_stream_state = {self.cursor_field: self.start_date}
+        return {self.cursor_field: max(latest_record.get(self.cursor_field, ""), current_stream_state.get(self.cursor_field, ""))}
+
+
+class CloseComActivitiesMixin(IncrementalCloseComStream):
```

Comment:
This isn't really a Mixin since we're not using it to achieve multiple inheritance. The structure looks like:
> CloseComStream
----> IncrementalCloseComStream
--------> CloseComActivitiesMixin
------------> CreatedActivities
------------> NoteActivities
------------> etc. etc.

In this sense, this class makes more sense as `CloseComActivitiesStream` imo.

Reply:
Thanks! Updated